pkg/typesystem/typesystem.go

```diff
@@ -667,163 +667,204 @@ func (t *TypeSystem) relationInvolvesExclusion(objectType, relation string, visi
+				// we entered a potential cycle, so keep on looking
 				continue
 			}
 
-			hasEntrypoint, _, err := hasEntrypoints(typedefs, assignableTypeName, assignableRelationName, assignableRelation.GetRewrite(), v)
+			relDetails, err := getRelationDetailsRecursive(typedefs, assignableTypeName, assignableRelationName, assignableRelation.GetRewrite(), v)
 			if err != nil {
-				return false, false, err
+				return nil, err
 			}
 
-			if hasEntrypoint {
-				return true, false, nil
+			if relDetails.hasEntrypoints {
+				v[typeName][relationName].hasEntrypoints = true
+				break
 			}
 		}
 
-		return false, false, nil
+		return v[typeName][relationName], nil
 	case *openfgav1.Userset_ComputedUserset:
 
 		computedRelationName := rw.ComputedUserset.GetRelation()
 		computedRelation, ok := typedefs[typeName][computedRelationName]
 		if !ok {
-			return false, false, fmt.Errorf("undefined type definition for '%s#%s'", typeName, computedRelationName)
+			return nil, fmt.Errorf("undefined type definition for '%s#%s'", typeName, computedRelationName)
 		}
 
-		if _, ok := v[typeName][computedRelationName]; ok {
-			return false, true, nil
+		if r, ok := v[typeName][computedRelationName]; ok {
+			// we entered a definitive cycle, so return
+			r.hasLoop = true
+			return r, nil
 		}
 
-		hasEntrypoint, loop, err := hasEntrypoints(typedefs, typeName, computedRelationName, computedRelation.GetRewrite(), v)
+		relDetails, err := getRelationDetailsRecursive(typedefs, typeName, computedRelationName, computedRelation.GetRewrite(), v)
 		if err != nil {
-			return false, false, err
+			return nil, err
 		}
 
-		return hasEntrypoint, loop, nil
+		return relDetails, nil
 	case *openfgav1.Userset_TupleToUserset:
 		tuplesetRelationName := rw.TupleToUserset.GetTupleset().GetRelation()
 		computedRelationName := rw.TupleToUserset.ComputedUserset.GetRelation()
 
 		tuplesetRelation, ok := typedefs[typeName][tuplesetRelationName]
 		if !ok {
-			return false, false, fmt.Errorf("undefined type definition for '%s#%s'", typeName, tuplesetRelationName)
+			return nil, fmt.Errorf("undefined type definition for '%s#%s'", typeName, tuplesetRelationName)
 		}
 
+		hasEntrypoints := false
+		// At least one type must have an entrypoint.
 		for _, assignableType := range tuplesetRelation.GetTypeInfo().GetDirectlyRelatedUserTypes() {
 			assignableTypeName := assignableType.GetType()
 
 			if assignableRelation, ok := typedefs[assignableTypeName][computedRelationName]; ok {
-				if _, ok := v[assignableTypeName][computedRelationName]; ok {
+				if r, ok := v[assignableTypeName][computedRelationName]; ok {
+					// We entered a potential cycle, so keep on looking.
+					// Note that, unlike the "This" case above, one TTU can add a parallel edge in the graph.
+					// For example: `define viewer: can_view from start and can_view from end`.
+					// So we need to remember if any of the edges led to "hasEntrypoints=true"
+					hasEntrypoints = hasEntrypoints || r.hasEntrypoints
```

Comment:
this fixes #1260. The model in that scenario is special because there is more than one edge between two nodes:

1. `state#can_view` to `transition#can_apply` via `can view from start`
2. `state#can_view` to `transition#can_apply` via `can view from end`